keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+      - downwardAPI:
+          items:
+          - path: namespace
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+```
+
+I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.
+
+### Risks and Mitigations
+
+**Scalability**: In the current design and draft implementation, Kubelet holds
+the private keys for each pod on the node in memory.  If Kubelet restarts, then
+the private keys are lost, and Kubelet will need to immediately generate new
+keys and request new certificates for each volume on the pod.  If Kubelet
+crashloops, this can generate high CertificateSigningRequest traffic on
+kube-apiserver.  This can potentially be mitigated by giving Kubelet the
+capability to read back the existing key from the projected volume on startup.
+
+## Design Details
+
+### X.509 PodIdentity Extension
+
+The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
+the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
+identity information into an X.509 certificate or certificate request, we can
+define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
+define a management policy for the PEN) and specify that it contains the
+following DER-encoded structure:
+
+```go
+type podIdentityASN1 struct {
+	Namespace          string `asn1:"utf8"`
+	ServiceAccountName string `asn1:"utf8"`
+	PodName            string `asn1:"utf8"`
+	PodUID             string `asn1:"utf8"`
+	NodeName           string `asn1:"utf8"`
+}
+```
+
+Certificates and CertificateRequests that embed this extension do not need to include any data in the standard X.509 Name or Subject Alternate Name fields.  The CertificateRequests generated by Kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.
+
+Utility functions for embedding and extracting this extension from X.509
+Certificate and CertificateRequest objects are defined in
+`k8s.io/component-helpers/kubernetesx509`.  They are intended to be used both by
+core Kubernetes components and third-party software that needs to interact with
+these certificates.
+
+### PodIdentity Extension Node Restriction
+
+The noderestriction admission plugin will check the consistency of PodIdentity extensions in CertificateSigningRequest objects created by `system:node:` identities.  The creation request will only be admitted if the PodIdentity extension refers to a pod that is currently running on the node that requested the certificate.
+
+This upholds the Kubernetes node isolation guarantee by ensuring that
+compromising a single node only results in compromises of workloads that have
+actually been scheduled onto that node.
+
+<<[UNRESOLVED]>>
+Is it necessary or desirable to forbid non-Kubelet identities from creating CertificateSigningRequest objects that embed a PodIdentity extension?
```

Comment:
Does anything prevent non-Kubelets from adding a Pod binding in TokenRequest, today? noderestriction doesn't appear to limit it, at least.

Reply:
I believe it is forbidden at some level.  But I guess we should just try it out.